deno_typescript/system_loader.js

```diff
@@ -4,96 +4,84 @@
 
 // @ts-nocheck
 /* eslint-disable */
-
-let System, __inst, __inst_s;
+let System, __ia, __i;
 
 (() => {
-  const mMap = new Map();
+  const r = new Map();
+
   System = {
     register(id, d, f) {
-      mMap.set(id, { id, d, f, exp: {} });
+      r.set(id, { d, f, exp: {} });
     }
   };
 
-  const gC = (data, main) => {
-    const { id } = data;
+  function gC(id, main) {
     return {
       id,
-      import: async id => mMap.get(id)?.exp,
+      import: async id => r.get(id)?.exp,
       meta: { url: id, main }
     };
-  };
+  }
 
-  const gE = ({ exp }) => {
+  function gE(exp) {
```

Comment:
There's a perf trick that is quite important here for large numbers of deep reexport statements. Basically checking the value of the export actually changed can avoid calling the setters. Since one export change can cause an export cascade up the dependency tree, whereby all the top-level export variables get reinitialized, checking that the value actually changed is quite important to avoid this cascade in reexporting scenarios since it can avoid the cascade before it starts. 

I forget the exact numbers, but it had a noticeable perf improvement for reexport trees when it was implemented.

Reply:
Thanks for the feedback.

In this implementation, I am trying to figure out where we would need to ever run the setters more than once.  Right now we follow the following process:

- `System.register()` of all modules.
- "main" module is passed, and all registered module factories are run (because we use `import.meta.main` we can't run the factories until we know the main module)
- The exports for the "main" module are requested which recursively triggers:
  - If the module setters haven't been run:
    - The exports of the dependencies of the module are gathered.
    - The setters are run.
    - The `execute()` is run.
  - The exports of the module are returned.

This means we only ever run the setters once.  In situations of circular dependencies, because we pull the setters off the module information, the reference to the exports are returned, skipping an infinite loop of trying to run the setters.

Because it is a "closed" environment, where all the potential modules are registered _before_ we try to resolve the exports of any module, I think we are potentially escaping the problem you mentioned above.

I know I have issues with dynamic `import()` (partly due to TypeScript not re-writing the module specifier, and therefore needing to implement some sort of resolution logic), but I still don't see a situation where I would need to run the setters more than once, except that in the circular tests I have, I think I am doing only namespace imports.  I guess I need to test named imports in a circular import situation.